docs/Operators.md

```diff
@@ -6307,13 +6307,13 @@ expect(node, inputs=[x], outputs=[y],
   
       graph body-net (
         %i[INT32, scalar]
-        %keepgoing[BOOL, scalar]
-        %b[INT32, scalar]
+        %keepgoing_in[BOOL, scalar]
+        %b_in[INT32, scalar]
       ) {
-        %my_local = Add(%a, %b)
-        %b_out = Sub(%a, %b)
+        %my_local = Add(%a, %b_in)
```

Comment:
It looks not smooth. The value of `b_in` may change from iteration to iteration, so adding an `Assignment` is closer to the sementic.

Reply:
I removed the changes to operator.md, and made the changes to the comments in defs.cc as suggested.
One issue  is that the "Loop" documentation does not specify the semantics for
zero-trip, i.e., if the loop body is not executred, what will be the outputs. I modified the equivalent C code to make the outputs hold  the values of the initial, and Scan-outputs will still be uninitialized.

Reply:
@gramalingam

Reply:
In the zero-trip case, I think it's correct to have the outputs hold the initial values for the loop carried dependencies  as I don't think they should be dropped due to having zero iterations. 

I'd describe the 'scan_outputs' as being empty rather than uninitialized. Outside of the scope of the C example, but the shape returned for a scan_output in the zero-trip case should have at least one dimension with a value of 0 in the first dimension as that represents the number of iterations, and downstream nodes will should be able to assume that dimension exists.

Reply:
However the C example seems flawed in that there's no good way to represent what is happening in the graph example. 

a and b are outer scope values, and there's a new scope for the loop implementation where b_in is provided as a parameter and b_out is a return value. there's no equivalent to that setup with a 'for' loop and that's where it gets confusing. 

Would the following be better?

    void loop_func()
    {
      /* User-defined code (enclosing scope) */
      int a = 3, b = 6;
      bool keepgoing = true; // Analogous to input cond
      /* End user-defined code */
      /* Implicitly-defined code */
      const int max_trip_count = 10; // Analogous to input M
      int user_defined_vals[]; // Imagine this is resizable
      /* End implicitly-defined code */

      /* initialize values that will be passed around during each iteration of the loop.
         also handles zero loop case */
      bool keepgoing_out = keepgoing;  // condition
      int b_out = b; // loop carried dependency
      
      for (int i=0; i < max_trip_count && keepgoing_out; ++i) {
        /* implicitly-defined code */  
        int b_in = b_out;  
        int keepgoing_in = keepgoing_out;

        /* User-defined code (loop body) */
        int my_local = a + b; // Reading values in the enclosing scope is fine
        b_out = a - b_in; // writes fine as b_out is the output value of the loop-carried dependency 
        keepgoing_out = my_local > b_out; 
        int user_defined_val = b_in + b_in;
        /* End user-defined code */

        /*implicitly-defined code */
        user_defined_vals[i] = user_defined_val;
      }

      // my_local = 123; // Can't do this. my_local was defined in the the body
      
      // These below values are live-out from the loop and therefore accessible
      // b_out; keepgoing_out; user_defined_vals; 
      // and user_defined_vals are concated output
    }